### ISSUE TITLE
Short open tag causes phpcbf to produce syntactically invalid file

### ISSUE THREAD

**Issue by paulschreiber** [squizlabs/PHP_CodeSniffer#2472]
Config
macOS 10.13.6
PHPCS 3.4.1
PHP 7.3.3

I came across a file like this:
```php
<?php
$orig_post = $post;
global $post;
$tags = wp_get_post_tags($post->ID);

if ($tags) :
$tag_ids = array();
foreach($tags as $individual_tag) $tag_ids[] = $individual_tag->name;
$args=array(
	'post_type' => 'brightcovevideos',
	'tag' => $tag_ids,
	'post__not_in' => array($post->ID),
	'posts_per_page'=>8 ,// Number of related posts that will be shown.
	'caller_get_posts'=>1
);
$my_query = new wp_query( $args );
if( $my_query->have_posts() ) :
?>


<?
endif;

else : ?>

<h1>No related posts. </h1>
<?php
endif;
?>
```

This file isn't completely valid — it has a short open tag (`<?`) before the `endif;`. 

When you run `phpcbf --standard=PSR2`, you get this:
```php
<?php
$orig_post = $post;
global $post;
$tags = wp_get_post_tags($post->ID);

if ($tags) {
    :
    $tag_ids = array();
}
foreach ($tags as $individual_tag) {
    $tag_ids[] = $individual_tag->name;
}
$args=array(
    'post_type' => 'brightcovevideos',
    'tag' => $tag_ids,
    'post__not_in' => array($post->ID),
    'posts_per_page'=>8 ,// Number of related posts that will be shown.
    'caller_get_posts'=>1
);
$my_query = new wp_query($args);
if ($my_query->have_posts()) :
    ?>


<?
endif;

else : ?>

<h1>No related posts. </h1>
    <?php
endif;
?>
```

This resulting file is not syntactically valid, even if you change the `<?` to `<?php`. Look at the `:` after `if ($tags) {`.

If you fix the short open tag first, you get this:
```php
<?php
$orig_post = $post;
global $post;
$tags = wp_get_post_tags($post->ID);

if ($tags) :
$tag_ids = array();
foreach($tags as $individual_tag) $tag_ids[] = $individual_tag->name;
$args=array(
	'post_type' => 'brightcovevideos',
	'tag' => $tag_ids,
	'post__not_in' => array($post->ID),
	'posts_per_page'=>8 ,// Number of related posts that will be shown.
	'caller_get_posts'=>1
);
$my_query = new wp_query( $args );
if( $my_query->have_posts() ) :
?>


<?php
endif;

else : ?>

<h1>No related posts. </h1>
<?php
endif;
?>
```

and after running `phpcbf --standard=PSR2`, you get a valid file:
```php
<?php
$orig_post = $post;
global $post;
$tags = wp_get_post_tags( $post->ID );

if ( $tags ) :
	$tag_ids = [];
	foreach ( $tags as $individual_tag ) {
		$tag_ids[] = $individual_tag->name;
	}
	$args     = [
		'post_type'        => 'brightcovevideos',
		'tag'              => $tag_ids,
		'post__not_in'     => [ $post->ID ],
		'posts_per_page'   => 8, // Number of related posts that will be shown.
		'caller_get_posts' => 1,
	];
	$my_query = new wp_query( $args );
	if ( $my_query->have_posts() ) :
		?>


		<?php
endif;

else :
	?>

<h1>No related posts. </h1>
	<?php
endif;
?>
```

**Comment by gsherwood** [comment]
The syntax error in your original file means that both PHP and PHPCS have tokenized it incorrectly. There was no way to match the first IF statement to a closing statement because the closing statement is not inside PHP code. PHPCBF did it's best to correct the file based on the structure it could see, but the file isn't valid, so the resulting file isn't going to be valid either.

I'm not sure what you are expecting PHPCBF to do in this case. Maybe you need to run the PHP linter over the files before trying to correct them? Or you can run PHPCS with the `Generic.PHP.Syntax` sniff, which can alert you to syntax errors in your files.

**Comment by paulschreiber** [comment]
Perhaps PHPCBF not try to rewrite files with syntax errors, since doing so will result in unpredictable/invalid results.

**Comment by gsherwood** [comment]
> Perhaps PHPCBF not try to rewrite files with syntax errors, since doing so will result in unpredictable/invalid results.

If it can auto-discover the PHP path, then I assume it could do this for you. It's not really how PHPCBF works though (it runs the sniffs and the sniffs do the fixing), so it would require some core changes to make it work.

I'll turn this issue into a feature request for that though.

**Comment by paulschreiber** [comment]
Thanks! I ran phpcbf on a project I inherited, and all seemed well. Then we went to production, and this page generated errors. Oops.
